### ISSUE TITLE
add threading.RWLock

### ISSUE THREAD

**Issue by 84401114-8e59-4056-83cb-632106c0b648** [python/cpython#53046]
BPO | [8800](https://bugs.python.org/issue8800)
--- | :---
Nosy | @jcea, @pitrou, @kristjanvalur, @tiran, @njsmith, @asvetlov, @ofek, @elarivie
Files | <li>[rwlock.patch](https://bugs.python.org/file17448/rwlock.patch "Uploaded as text/plain at 2010-05-24.09:55:37 by @kristjanvalur")</li><li>[Added-ShrdExclLock-to-threading-and-multiprocessing.patch](https://bugs.python.org/file27350/Added-ShrdExclLock-to-threading-and-multiprocessing.patch "Uploaded as text/plain at 2012-09-30.02:01:39 by Sebastian.Noack")</li><li>[sharablelock.patch](https://bugs.python.org/file27359/sharablelock.patch "Uploaded as text/plain at 2012-09-30.17:19:05 by @kristjanvalur")</li><li>[Added-ShrdExclLock-to-threading-and-multiprocessing-2.patch](https://bugs.python.org/file27363/Added-ShrdExclLock-to-threading-and-multiprocessing-2.patch "Uploaded as text/plain at 2012-09-30.18:40:15 by Sebastian.Noack")</li><li>[rwlock.patch](https://bugs.python.org/file27385/rwlock.patch "Uploaded as text/plain at 2012-10-02.11:37:48 by @kristjanvalur")</li><li>[rwlock.patch](https://bugs.python.org/file27412/rwlock.patch "Uploaded as text/plain at 2012-10-04.10:29:43 by @kristjanvalur")</li><li>[rwlock-sbt.patch](https://bugs.python.org/file27422/rwlock-sbt.patch "Uploaded as text/plain at 2012-10-04.17:24:23 by sbt")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2010-05-24.09:55:40.864>
labels = ['type-feature', 'library']
title = 'add threading.RWLock'
updated_at = <Date 2020-01-15.09:20:07.038>
user = 'https://github.com/kristjanvalur'
```

bugs.python.org fields:
```python
activity = <Date 2020-01-15.09:20:07.038>
actor = 'asvetlov'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2010-05-24.09:55:40.864>
creator = 'kristjan.jonsson'
dependencies = []
files = ['17448', '27350', '27359', '27363', '27385', '27412', '27422']
hgrepos = []
issue_num = 8800
keywords = ['patch', 'needs review']
message_count = 63.0
messages = ['106350', '106372', '106373', '106376', '106377', '106386', '168077', '168149', '171567', '171568', '171599', '171600', '171626', '171639', '171653', '171659', '171667', '171669', '171674', '171693', '171695', '171696', '171697', '171698', '171699', '171700', '171703', '171708', '171709', '171710', '171713', '171714', '171716', '171717', '171718', '171721', '171780', '171781', '171782', '171783', '171785', '171786', '171788', '171789', '171790', '171792', '171793', '171877', '171883', '171891', '171914', '171915', '171930', '171975', '171979', '172064', '172071', '172074', '274765', '274795', '287745', '360026', '360031']
nosy_count = 15.0
nosy_names = ['jcea', 'pitrou', 'kristjan.jonsson', 'christian.heimes', 'jyasskin', 'njs', 'asvetlov', 'neologix', 'vrutsky', 'sbt', 'mklauber', 'Sebastian.Noack', 'dan.oreilly', 'Ofekmeister', 'elarivie']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue8800'
versions = []
```

</p></details>

**Comment by njsmith** [comment]
Side note, in case anyone else finds themselves looking for the MSDN magazine article referenced above: it appears to be in the June 2007 issue, titled "CONCURRENCY: Synchronization Primitives New To Windows Vista". 

For the moment at least, this link seems to work: http://download.microsoft.com/download/3/A/7/3A7FA450-1F33-41F7-9E6D-3AA95B5A6AEA/MSDNMagazineJune2007en-us.chm

**Comment by zklaus** [comment]
Is there still interest in having an RWLock in python itself? I personally think it would be an interesting and important addition.

It seems this issue opens with a patch and quickly dives into implementation questions, but perhaps it would be good to have a general consensus on whether or not this *should* be added?

**Comment by pitrou** [comment]
I do think a RWLock (or SELock, whetever the name :-)) would be useful.

@gpshead What do you think?

**Comment by pitrou** [comment]
That said, and for the record, I probably wouldn't review such a PR as I'm not active anymore.

**Comment by lrlunin** [comment]
What is an official status of this feature request? Would a sufficient solution/implementation added to cpython if there is any?

**Comment by gpshead** [comment]
Given that https://pypi.org/project/readerwriterlock/ exists and there hasn't been a high demand for this, I suggest not adding it to the standard library. We provide the basics, things like that build upon them.

I did skim over the [`rwlock-sbt.patch`](https://bugs.python.org/file27422/rwlock-sbt.patch) linked above (which appeared to be the most recent iteration of the 11y old proposals) and it isn't necessarily a bad thing, but a reader/writer shared/exclusive lock implementation is a complicated thing to maintain.

If we were to add one... I think we should really not be reinventing the wheel ourselves, in pure Python. It'd be good to wrap the underlying platform's implementation such as pthread_rwlock on posix and presumably something equivalent available on Windows.  Those will perform better.

Similar to how we had a poorly performing pure python RLock implementation in the past until https://github.com/python/cpython/issues/47251 (bpo-3001) fixed it in ~3.2.

**Comment by rafalkrupinski** [comment]
> We provide the basics, things like that build upon them.

How come toml is basics and RWLock is not?
